File: src/gui/dialogs/import_dialog.py
"""
Cross-Project Import Dialog for NC-KTV
Allows importing components from other .nctv files
"""

from PyQt6.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QLabel, 
    QPushButton, QCheckBox, QGroupBox, QFileDialog,
    QMessageBox, QLineEdit
)
from PyQt6.QtCore import Qt
from pathlib import Path
from typing import Optional

from core.project import Project


class ImportDialog(QDialog):
    """Dialog for importing components from another project"""
    
    def __init__(self, current_project: Project, parent=None):
        super().__init__(parent)
        self.current_project = current_project
        self.import_project: Optional[Project] = None
        
        self.setWindowTitle("Import from Project")
        self.setModal(True)
        self.resize(500, 400)
        
        self._init_ui()
    
    def _init_ui(self):
        """Initialize UI"""
        layout = QVBoxLayout(self)
        
        # File selection
        file_group = QGroupBox("Source Project")
        file_layout = QHBoxLayout()
        
        self.file_path_edit = QLineEdit()
        self.file_path_edit.setReadOnly(True)
        self.file_path_edit.setPlaceholderText("No file selected...")
        file_layout.addWidget(self.file_path_edit)
        
        btn_browse = QPushButton("📁 Browse...")
        btn_browse.clicked.connect(self._browse_file)
        file_layout.addWidget(btn_browse)
        
        file_group.setLayout(file_layout)
        layout.addWidget(file_group)
        
        # Import options
        options_group = QGroupBox("Components to Import")
        options_layout = QVBoxLayout()
        
        self.cb_lyrics = QCheckBox("Lyrics (text and timestamps)")
        self.cb_lyrics.setChecked(True)
        options_layout.addWidget(self.cb_lyrics)
        
        self.cb_audio = QCheckBox("Audio files (instrumental, vocals)")
        options_layout.addWidget(self.cb_audio)
        
        self.cb_metadata = QCheckBox("Metadata (settings, dates)")
        options_layout.addWidget(self.cb_metadata)
        
        options_layout.addStretch()
        options_group.setLayout(options_layout)
        layout.addWidget(options_group)
        
        # Merge strategy
        merge_group = QGroupBox("Merge Strategy")
        merge_layout = QVBoxLayout()
        
        info_label = QLabel(
            "⚠️ Importing will overwrite existing data for selected components.\n"
            "This action cannot be undone."
        )
        info_label.setStyleSheet("color: #FF9800; padding: 10px;")
        info_label.setWordWrap(True)
        merge_layout.addWidget(info_label)
        
        merge_group.setLayout(merge_layout)
        layout.addWidget(merge_group)
        
        # Buttons
        button_layout = QHBoxLayout()
        button_layout.addStretch()
        
        btn_cancel = QPushButton("Cancel")
        btn_cancel.clicked.connect(self.reject)
        button_layout.addWidget(btn_cancel)
        
        self.btn_import = QPushButton("✅ Import")
        self.btn_import.clicked.connect(self._perform_import)
        self.btn_import.setEnabled(False)
        self.btn_import.setStyleSheet("background-color: #4CAF50; color: white; font-weight: bold;")
        button_layout.addWidget(self.btn_import)
        
        layout.addLayout(button_layout)
    
    def _browse_file(self):
        """Open file browser to select project"""
        file_path, _ = QFileDialog.getOpenFileName(
            self,
            "Select Project to Import",
            "",
            "NC-KTV Project (*.nctv)"
        )
        
        if file_path:
            try:
                # Load project
                self.import_project = Project.load(Path(file_path))
                self.file_path_edit.setText(file_path)
                self.btn_import.setEnabled(True)
                
                QMessageBox.information(
                    self,
                    "Success",
                    f"Loaded project: {self.import_project.project_name}"
                )
            except Exception as e:
                QMessageBox.critical(
                    self,
                    "Error",
                    f"Failed to load project:\n{e}"
                )
                self.import_project = None
                self.file_path_edit.clear()
                self.btn_import.setEnabled(False)
    
    def _perform_import(self):
        """Import selected components"""
        if not self.import_project:
            return
        
        # Confirm import
        reply = QMessageBox.question(
            self,
            "Confirm Import",
            "Are you sure you want to import the selected components?\n"
            "This will overwrite existing data.",
            QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No
        )
        
        if reply != QMessageBox.StandardButton.Yes:
            return
        
        try:
            # Import lyrics
            if self.cb_lyrics.isChecked():
                self.current_project.lyrics = self.import_project.lyrics
            
            # Import audio files
            if self.cb_audio.isChecked():
                self.current_project.instrumental_file = self.import_project.instrumental_file
                self.current_project.vocals_file = self.import_project.vocals_file
            
            # Import metadata
            if self.cb_metadata.isChecked():
                self.current_project.settings = self.import_project.settings
                # Note: We don't overwrite dates to preserve current project's timeline
            
            QMessageBox.information(
                self,
                "Success",
                "Components imported successfully!"
            )
            
            self.accept()
            
        except Exception as e:
            QMessageBox.critical(
                self,
                "Error",
                f"Import failed:\n{e}"
            )

File: cpp/src/gui/dialogs/import_dialog.cpp
/*
 * NC-KTV GUI — Import Dialog Implementation
 */

#include "import_dialog.h"
#include "lyrics_search_dialog.h"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFileDialog>
#include <QFileInfo>
#include <QFile>
#include <QMessageBox>
#include <QGroupBox>

namespace ncktv {

ImportDialog::ImportDialog(QWidget* parent)
    : QDialog(parent)
{
    setWindowTitle("Import Subtitles");
    setMinimumSize(560, 440);
    setModal(true);
    setupUi();
    applyTheme();
}

void ImportDialog::setupUi() {
    auto* mainLayout = new QVBoxLayout(this);
    mainLayout->setSpacing(14);
    mainLayout->setContentsMargins(24, 20, 24, 20);

    // ── Title ────────────────────────────────────────────────────────────
    auto* titleLabel = new QLabel("📁  Import Subtitles / Lyrics", this);
    titleLabel->setStyleSheet("font-size:18px; font-weight:bold; color:#e0e0e0; border:none;");
    mainLayout->addWidget(titleLabel);

    auto* descLabel = new QLabel("Select a subtitle or lyrics file to import. Supports LRC, SRT, TXT, VTT, ASS, and Whisper JSON.", this);
    descLabel->setWordWrap(true);
    descLabel->setStyleSheet("color:#888; font-size:12px; border:none;");
    mainLayout->addWidget(descLabel);

    // ── File selector ────────────────────────────────────────────────────
    auto* fileGroup = new QGroupBox("Source File", this);
    auto* fileLayout = new QVBoxLayout(fileGroup);

    auto* browseLayout = new QHBoxLayout();
    m_pathEdit = new QLineEdit(this);
    m_pathEdit->setPlaceholderText("Select a subtitle file...");
    m_pathEdit->setReadOnly(true);
    m_browseBtn = new QPushButton("Browse...", this);
    m_browseBtn->setFixedWidth(90);
    m_searchBtn = new QPushButton("🔍 Search Online", this);
    m_searchBtn->setFixedWidth(120);
    m_searchBtn->setStyleSheet("background: #1e293b; color: #94a3b8; font-weight: bold; border-color: #334155;");
    
    browseLayout->addWidget(m_pathEdit, 1);
    browseLayout->addWidget(m_browseBtn);
    browseLayout->addWidget(m_searchBtn);
    fileLayout->addLayout(browseLayout);

    auto* infoLayout = new QHBoxLayout();
    m_formatLabel = new QLabel("Format: —", this);
    m_formatLabel->setStyleSheet("color:#4f46e5; font-weight:bold; border:none;");
    m_lineCountLabel = new QLabel("", this);
    m_lineCountLabel->setStyleSheet("color:#888; border:none;");
    infoLayout->addWidget(m_formatLabel);
    infoLayout->addStretch();
    infoLayout->addWidget(m_lineCountLabel);
    fileLayout->addLayout(infoLayout);

    mainLayout->addWidget(fileGroup);

    // ── Preview ──────────────────────────────────────────────────────────
    auto* previewGroup = new QGroupBox("Preview", this);
    auto* previewLayout = new QVBoxLayout(previewGroup);
    m_previewText = new QTextEdit(this);
    m_previewText->setReadOnly(true);
    m_previewText->setPlaceholderText("File preview will appear here...");
    m_previewText->setMaximumHeight(160);
    previewLayout->addWidget(m_previewText);
    mainLayout->addWidget(previewGroup, 1);

    // ── Buttons ──────────────────────────────────────────────────────────
    auto* buttonLayout = new QHBoxLayout();
    buttonLayout->addStretch();

    m_cancelBtn = new QPushButton("Cancel", this);
    m_cancelBtn->setFixedWidth(100);
    buttonLayout->addWidget(m_cancelBtn);

    m_importBtn = new QPushButton("Import", this);
    m_importBtn->setFixedWidth(120);
    m_importBtn->setDefault(true);
    m_importBtn->setEnabled(false);
    m_importBtn->setStyleSheet("background:#4f46e5; color:white; font-weight:bold; border:none; padding:8px 16px; border-radius:4px;");
    buttonLayout->addWidget(m_importBtn);

    mainLayout->addLayout(buttonLayout);

    // ── Connections ──────────────────────────────────────────────────────
    connect(m_browseBtn, &QPushButton::clicked, this, &ImportDialog::browseFile);
    connect(m_searchBtn, &QPushButton::clicked, this, &ImportDialog::searchOnline);
    connect(m_importBtn, &QPushButton::clicked, this, &ImportDialog::onAccepted);
    connect(m_cancelBtn, &QPushButton::clicked, this, &QDialog::reject);
}

void ImportDialog::applyTheme() {
    setStyleSheet(R"(
        QDialog { background:#1e1e1e; color:#d4d4d4; }
        QGroupBox { border:1px solid #3f3f46; border-radius:6px; padding-top:16px; margin-top:8px; font-weight:bold; color:#ccc; }
        QGroupBox::title { subcontrol-origin:margin; left:12px; padding:0 6px; }
        QLineEdit { background:#2d2d30; border:1px solid #3f3f46; border-radius:4px; padding:6px 8px; color:#fff; }
        QTextEdit { background:#2d2d30; border:1px solid #3f3f46; border-radius:4px; padding:6px; color:#d4d4d4; font-family:"Consolas","Courier New",monospace; font-size:12px; }
        QPushButton { background:#333; border:1px solid #555; border-radius:4px; padding:6px 14px; color:#ccc; }
        QPushButton:hover { background:#404040; }
        QPushButton:disabled { background:#2a2a2a; color:#555; }
        QLabel { color:#ccc; border:none; }
    )");
}

void ImportDialog::browseFile() {
    QString filter = "Subtitle Files (*.lrc *.srt *.txt *.json *.vtt *.ass *.ssa *.ttml);;Lyric (*.lrc);;SubRip (*.srt);;Text (*.txt);;Whisper JSON (*.json);;All Files (*)";
    QString path = QFileDialog::getOpenFileName(this, "Select Subtitle File", "", filter);
    if (!path.isEmpty()) {
        onFileSelected(path);
    }
}

void ImportDialog::searchOnline() {
    LyricsSearchDialog dialog(this);
    
    // If we have a filename already, try to use it as initial search
    if (!m_pathEdit->text().isEmpty()) {
        QFileInfo fi(m_pathEdit->text());
        dialog.setInitialSearch(fi.baseName());
    }

    if (dialog.exec() == QDialog::Accepted) {
        QString content = dialog.syncedLyrics();
        if (content.isEmpty()) content = dialog.plainLyrics();

        if (content.isEmpty()) return;

        // Save to temporary file
        QString tempPath = QDir::tempPath() + "/ncktv_online_lyrics.lrc";
        QFile file(tempPath);
        if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
            file.write(content.toUtf8());
            file.close();
            onFileSelected(tempPath);
            
            // Mark it as online
            m_formatLabel->setText("Format: LRC (Online via LRCLIB)");
        }
    }
}

void ImportDialog::onFileSelected(const QString& path) {
    m_pathEdit->setText(path);
    detectFormat(path);

    // Load preview (first 30 lines)
    QFile file(path);
    if (file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        QString content = QString::fromUtf8(file.readAll());
        QStringList lines = content.split('\n');
        int totalLines = lines.size();
        m_lineCountLabel->setText(QString("%1 lines").arg(totalLines));

        // Show first 30 lines
        QStringList preview = lines.mid(0, 30);
        m_previewText->setPlainText(preview.join('\n'));
        if (totalLines > 30) {
            m_previewText->append(QString("\n... (%1 more lines)").arg(totalLines - 30));
        }
        m_importBtn->setEnabled(true);
    } else {
        m_previewText->setPlainText("(Could not read file)");
        m_importBtn->setEnabled(false);
    }
}

void ImportDialog::detectFormat(const QString& path) {
    QFileInfo fi(path);
    QString ext = fi.suffix().toLower();

    if (ext == "lrc")       m_detectedFormat = "LRC";
    else if (ext == "srt")  m_detectedFormat = "SRT";
    else if (ext == "vtt")  m_detectedFormat = "VTT";
    else if (ext == "ass")  m_detectedFormat = "ASS";
    else if (ext == "ssa")  m_detectedFormat = "SSA";
    else if (ext == "ttml") m_detectedFormat = "TTML";
    else if (ext == "json") m_detectedFormat = "Whisper JSON";
    else if (ext == "txt")  m_detectedFormat = "Plain Text";
    else                    m_detectedFormat = "Unknown";

    m_formatLabel->setText("Format: " + m_detectedFormat);
}

void ImportDialog::onAccepted() {
    if (m_pathEdit->text().isEmpty()) {
        QMessageBox::warning(this, "No File", "Please select a file to import.");
        return;
    }
    accept();
}

QString ImportDialog::filePath()       const { return m_pathEdit->text(); }
QString ImportDialog::detectedFormat() const { return m_detectedFormat; }

} // namespace ncktv
